confgenerator/agentmetrics.go

```diff
@@ -28,11 +27,192 @@ import (
 // It is never referenced in the config file, and instead is forcibly added in confgenerator.go.
 // Therefore, it does not need to implement any interfaces.
 type AgentSelfMetrics struct {
-	Version string
-	Port    int
+	MetricsVersionLabel string
+	LoggingVersionLabel string
+	FluentBitPort       int
+	OtelPort            int
+	OtelLoggingEnabled  bool
+	OtelRuntimeDir      string
 }
 
-func (r AgentSelfMetrics) MetricsSubmodulePipeline() otel.ReceiverPipeline {
+var grpcToHTTPStatus = map[string]string{
+	"OK":                  "200",
+	"INVALID_ARGUMENT":    "400",
+	"FAILED_PRECONDITION": "400",
+	"OUT_OF_RANGE":        "400",
+	"UNAUTHENTICATED":     "401",
+	"PERMISSION_DENIED":   "403",
+	"NOT_FOUND":           "404",
+	"ALREADY_EXISTS":      "409",
+	"ABORTED":             "409",
+	"RESOURCE_EXHAUSTED":  "429",
+	"CANCELLED":           "499",
+	"UNKNOWN":             "500",
+	"INTERNAL":            "500",
+	"DATA_LOSS":           "500",
+	"UNIMPLEMENTED":       "501",
+	"UNAVAILABLE":         "503",
+	"DEADLINE_EXCEEDED":   "504",
+}
+
+func (r AgentSelfMetrics) otelPipelineProcessorsWithOtelLogging() map[string][]otel.Component {
+	return map[string][]otel.Component{"metrics": {
+		otel.MetricsFilter(
+			"include",
+			"strict",
+			"otelcol_process_uptime",
+			"otelcol_process_memory_rss",
+			"otelcol_exporter_sent_log_records",
+			"otelcol_exporter_send_failed_log_records",
+			"grpc.client.attempt.duration",
+			"googlecloudmonitoring/point_count",
+		),
+		otel.Transform("metric", "metric",
+			[]ottl.Statement{
+				// Make two copies of request duration for logging and monitoring
+				ottl.CopyMetric("grpc.client.attempt.duration.logging", `grpc.client.attempt.duration`),
+				ottl.CopyMetric("grpc.client.attempt.duration.monitoring", `grpc.client.attempt.duration`),
+				// Extract count from histogram metrics
+				ottl.ExtractCountMetric(true, "grpc.client.attempt.duration.logging"),
+				ottl.ExtractCountMetric(true, "grpc.client.attempt.duration.monitoring"),
+			},
+		),
+		otel.MetricsOTTLFilter([]string{}, []string{
+			// Filter out histogram datapoints where the grpc.target is not related.
+			`metric.name == "grpc.client.attempt.duration.logging_count" and (not IsMatch(datapoint.attributes["grpc.target"], "logging.googleapis"))`,
+			`metric.name == "grpc.client.attempt.duration.monitoring_count" and (not IsMatch(datapoint.attributes["grpc.target"], "monitoring.googleapis"))`,
+		}),
+		otel.MetricsFilter(
+			"include",
+			"strict",
+			"otelcol_process_uptime",
+			"otelcol_process_memory_rss",
+			"otelcol_exporter_sent_log_records",
+			"otelcol_exporter_send_failed_log_records",
+			"grpc.client.attempt.duration.logging_count",
+			"grpc.client.attempt.duration.monitoring_count",
+			"googlecloudmonitoring/point_count",
+		),
+		otel.MetricsTransform(
+			otel.DuplicateMetric("otelcol_exporter_send_failed_log_records", "agent/log_entry_retry_count",
```

Comment:
Do you have anything to prevent this from colliding with the stream that comes from fluent-bit?

Reply:
I don't think we should do that - if the process is still running, we should still be exporting its metrics. Only when we actually remove fluent-bit entirely can we disable the fluent-bit metric scraping.